File: WitchcraftApp/src/main/java/com/tarbadev/witchcraft/recipes/persistence/entity/RecipeEntity.java
package com.tarbadev.witchcraft.recipes.persistence.entity;

import com.tarbadev.witchcraft.recipes.domain.entity.Recipe;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import javax.persistence.*;
import java.util.List;
import java.util.stream.Collectors;

@Builder
@Data
@AllArgsConstructor
@NoArgsConstructor
@Entity(name = "recipes")
public class RecipeEntity {
  @Id
  @GeneratedValue(strategy = GenerationType.AUTO)
  private Integer id;
  private String originUrl;
  private String name;
  private String imgUrl;
  @Builder.Default
  private Boolean favorite = false;
  @OneToMany(cascade = CascadeType.ALL)
  @JoinColumn(name = "recipe_id")
  private List<IngredientEntity> ingredients;
  @OneToMany(cascade = CascadeType.ALL)
  @JoinColumn(name = "recipe_id")
  private List<StepEntity> steps;

  public RecipeEntity(Recipe recipe) {
    id = recipe.getId();
    originUrl = recipe.getOriginUrl();
    name = recipe.getName();
    imgUrl = recipe.getImgUrl();
    favorite = recipe.getFavorite();
    ingredients = recipe.getIngredients().stream().map(IngredientEntity::new).collect(Collectors.toList());
    steps = recipe.getSteps().stream().map(StepEntity::new).collect(Collectors.toList());
  }

  public Recipe recipe() {
    return Recipe.builder()
        .id(id)
        .name(name.toLowerCase())
        .url("/recipes/" + id)
        .originUrl(originUrl)
        .imgUrl(imgUrl)
        .favorite(favorite)
        .ingredients(ingredients.stream().map(IngredientEntity::ingredient).collect(Collectors.toList()))
        .steps(steps.stream().map(StepEntity::step).collect(Collectors.toList()))
        .build();
  }
}

File: WitchcraftApp/src/main/java/com/tarbadev/witchcraft/recipes/persistence/entity/RecipeEntity.kt
package com.tarbadev.witchcraft.recipes.persistence.entity

import com.tarbadev.witchcraft.recipes.domain.entity.Recipe
import javax.persistence.*

@Entity(name = "recipes")
data class RecipeEntity(
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    val id: Int = 0,
    val originUrl: String = "",
    val name: String = "",
    val imgUrl: String = "",
    var favorite: Boolean = false,
    @OneToMany(cascade = [CascadeType.ALL])
    @JoinColumn(name = "recipe_id")
    @OrderBy("name ASC")
    val ingredients: List<IngredientEntity> = emptyList(),
    @OneToMany(cascade = [CascadeType.ALL])
    @JoinColumn(name = "recipe_id")
    val steps: Set<StepEntity> = emptySet()
) {
    constructor(recipe: Recipe) : this(
        id = recipe.id,
        originUrl = recipe.originUrl,
        name = recipe.name,
        imgUrl = recipe.imgUrl,
        favorite = recipe.favorite,
        ingredients = recipe.ingredients.map { IngredientEntity(it) },
        steps = recipe.steps.map { StepEntity(it) }.toSet()
    )

    fun recipe(): Recipe {
        return Recipe(
            id = id,
            name = name.toLowerCase(),
            url = "/recipes/$id",
            originUrl = originUrl,
            imgUrl = imgUrl,
            favorite = favorite,
            ingredients = ingredients.map { it.ingredient() },
            steps = steps.map { it.step() }
        )
    }
}
